File: Depth-First Search/Binary tree/Searching/Medium/LET_no230_Kth Smallest Element in a BST.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        stack = []    # Space: O(h) where h is the tree's height
        while root:    # Time: O(h)
            stack.append(root)
            root = root.left
            
        res = None
        for _ in range(k):    # Time: O(k)
            cur = stack.pop()
            res = cur.val
            if cur.right:
                cur = cur.right
                while cur:
                    stack.append(cur)
                    cur = cur.left
        
        return res

File: Depth-First Search/Binary tree/Searching/Medium/LET_no230_Kth Smallest Element in a BST.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        vector<TreeNode*> stack;
        while (root) {
            stack.push_back(root);
            root = root->left;
        }

        int res;
        TreeNode* cur;
        while (k-- > 0) {
            cur = stack.back();
            stack.pop_back();
            res = cur->val;
            if (cur->right) {
                cur = cur->right;
                while (cur) {
                    stack.push_back(cur);
                    cur = cur->left;
                }
            }
        }

        return res;
    }
};